File: java/src/jan/week1/boj/가장긴증가하는부분수열5_박재환.java
package jan.week1.boj;

import java.util.*;
import java.io.*;

public class 가장긴증가하는부분수열5_박재환 {
    static BufferedReader br;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        sb = new StringBuilder();
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
        System.out.println(sb);
    }

    static StringTokenizer st;
    static int n;
    static int[] arr;
    static int[] arrLoc;
    static List<Integer> lisList;
    static void init() throws IOException {
        lisList = new ArrayList<>();
        n = Integer.parseInt(br.readLine().trim());
        arr = new int[n];
        arrLoc = new int[n];
        st = new StringTokenizer(br.readLine().trim());
        for(int i=0; i<n; i++) arr[i] = Integer.parseInt(st.nextToken());
        getLis();

        int lisLen = lisList.size();
        sb.append(lisLen).append('\n');
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> Integer.compare(a, b));
        for(int i=n-1; i>-1 && lisLen > 0; i--) {
            if(arrLoc[i] == lisLen) {
                lisLen--;
                pq.offer(arr[i]);
            }
        }
        while(!pq.isEmpty()) sb.append(pq.poll()).append(' ');
    }

    static void getLis() {
        for(int i=0; i<n; i++) {
            int target = arr[i];
            if(lisList.isEmpty() || lisList.get(lisList.size()-1) < target) {
                lisList.add(target);
                arrLoc[i] = lisList.size();
                continue;
            }

            int insertIdx = insertIdx(target);
            lisList.set(insertIdx, target);
            arrLoc[i] = insertIdx+1;
        }
    }

    static int insertIdx(int target) {
        int l = 0, r = lisList.size()-1;

        while(l < r) {
            int mid = l + (r-l)/2;

            if(lisList.get(mid) >= target) r = mid;
            else l = mid+1;
        }

        return l;
    }
}

File: python/src/jan/week1/boj/가장긴증가하는부분수열5_박재환.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
arr_idx = [0] * n
lis_list = []

def lis():
    from collections import deque

    for i in range(n):
        if len(lis_list) == 0 or lis_list[-1] < arr[i]:
            lis_list.append(arr[i])
            arr_idx[i] = len(lis_list)
            continue
        insert_idx = get_insert_idx(arr[i])
        arr_idx[i] = insert_idx+1
        lis_list[insert_idx] = arr[i]

    print(len(lis_list))
    dq = deque()
    lis_len = len(lis_list)
    for i in range(n-1, -1, -1):
        if lis_len == arr_idx[i]:
            dq.appendleft(arr[i])
            lis_len-=1

    print(*dq)

def get_insert_idx(target):
    l = 0
    r = len(lis_list) - 1

    while(l < r):
        mid = l + (r-l)//2
        if lis_list[mid] >= target:
            r = mid
        else:
            l = mid + 1
    return l

if __name__ == '__main__':
    lis()